VisualStudioPlugin/SolutionExplorerMenuCommand.cs

```diff
@@ -163,47 +156,121 @@ private void SaveToQuantConnectCallback(object sender, EventArgs e)
-        private void UploadFilesToServer(int selectedProjectId, List<SelectedItem> files)
+        /// <summary>
+        /// Uploads a list of files to a specific project at QuantConnect
+        /// </summary>
+        /// <param name="selectedProjectId">Target project Id</param>
+        /// <param name="files">List of files to upload</param>
+        /// <returns>Returns false if any file failed to be uploaded</returns>
+        private bool UploadFilesToServer(int selectedProjectId, IEnumerable<SelectedItem> files)
         {
+            VsUtils.DisplayInStatusBar(_serviceProvider, "Uploading files to server ...");
             var api = AuthorizationManager.GetInstance().GetApi();
+            // Counters to keep track of files uploaded or not
+            var filesUploaded = 0;
+            var filesNotUploaded = 0;
             foreach (var file in files)
             {
                 api.DeleteProjectFile(selectedProjectId, file.FileName);
                 var fileContent = File.ReadAllText(file.FilePath);
-                api.AddProjectFile(selectedProjectId, file.FileName, fileContent);
+                var response = api.AddProjectFile(selectedProjectId, file.FileName, fileContent);
+                if (response.Success)
+                {
+                    filesUploaded++;
+                }
+                else
+                {
+                    VSActivityLog.Error("Failed to add project file " + file.FileName);
+                    filesNotUploaded++;
+                }
             }
+            // Update Status bar accordingly based on counters
+            var message = "Files upload complete";
+            message += (filesUploaded != 0) ? ". Uploaded " + filesUploaded : "";
+            message += (filesNotUploaded != 0) ? ". Failed to upload " + filesNotUploaded : "";
+            VsUtils.DisplayInStatusBar(_serviceProvider, message);
+
+            // Return false if any file failed to be uploaded
+            var res = filesNotUploaded == 0;
+            if (!res)
+            {
+                VsUtils.ShowMessageBox(_serviceProvider, "Upload Files Failed", message);
+            }
+            return res;
         }
 
-        private Api.Compile CompileProjectOnServer(int projectId)
+        /// <summary>
+        /// Compiles specific projectId at QuantConnect
+        /// </summary>
+        /// <param name="projectId">Target project Id</param>
+        /// <returns>Returns empty if failed else compile Id</returns>
+        private string CompileProjectOnServer(int projectId)
         {
+            VsUtils.DisplayInStatusBar(_serviceProvider, "Compiling project ...");
             var api = AuthorizationManager.GetInstance().GetApi();
             var compileStatus = api.CreateCompile(projectId);
             var compileId = compileStatus.CompileId;
             while (compileStatus.State == Api.CompileState.InQueue)
             {
-                Thread.Sleep(5000);
+                Thread.Sleep(2500);
```

Comment:
I see this hanging when you deploy for backtesting -- so it might be synchronous with the UI thread. Please use Tasks.ContinueWith to execute code blocks one after the other asynchronously.

Reply:
Fixed. Both 'Send for backtesting' and 'Save to QuantConnect' are now non UI blocking